### PR TITLE
cppcheck29

### PR DESCRIPTION
**Pull request by MartinDelille**
- cppcheck/2.9
- Conan v2 migration

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ ] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by jgsogo** [comment]
Although the CI didn't report anything (it was broken this morning), the linter already contains an error about `replace_in_file` that need to be fixed. After the fix (and push) CI should trigger as usual.

**Comment by uilianries** [comment]
@MartinDelille you need to remove `base_path` for all patches in conandata.yml

**Comment by MartinDelille** [comment]
No idea what is going wrong now...

**Comment by uilianries** [comment]
The [GMP](https://conan.io/center/gmp) is missing, but is not listed on dependencies. Maybe, it has been introduced now and you need to add as a new dependency.

**Comment by MartinDelille** [comment]
If I add `gmp/6.2.1` to requirements,, I have the following error on MacOS:

```
$  conan create cppcheck/all cppcheck/2.9@
Exporting package recipe
cppcheck/2.9 exports: File 'conandata.yml' found. Exporting it...
cppcheck/2.9 exports: Copied 1 '.yml' file: conandata.yml
cppcheck/2.9: Calling export_sources()
cppcheck/2.9: The stored package has not changed
cppcheck/2.9: Exported revision: 0501d80fe05c3be967bc4f1f7c9a7d20
Configuration:
[settings]
arch=x86_64
arch_build=x86_64
build_type=Debug
compiler=apple-clang
compiler.libcxx=libc++
compiler.version=13.1
os=Macos
os.version=10.14
os_build=Macos
[options]
[build_requires]
[env]

cppcheck/2.9: Forced build from source
ERROR: At least two recipes provides the same functionality:
 - 'gmp' provided by 'gmp/6.2.1', 'mpir/3.0.0'
```

**Comment by jgsogo** [comment]
According to docs, `mpir` is basically a replacement of `gmp` (https://www.swmath.org/software/4700):

> MPIR is an open source multiprecision integer (bignum) library forked from the GMP (GNU Multi Precision) project. It consists of much code from past GMP releases, in combination with much original contributed code

File: recipes/cppcheck/all/conanfile.py [hunk 1]
```diff
@@ -12,59 +13,54 @@ class CppcheckConan(ConanFile):
...
                               "RUNTIME DESTINATION ${CMAKE_INSTALL_FULL_BINDIR}",
                               "DESTINATION ${CMAKE_INSTALL_FULL_BINDIR}")
 
+    def export_sources(self):
+        for p in self.conan_data.get("patches", {}).get(self.version, []):
+            copy(self, p["patch_file"], self.recipe_folder, self.export_sources_folder)
+
     def requirements(self):
         if self.options.with_z3:
             self.requires("z3/4.8.8")
```
Comment (ericriff) [inline recipes/cppcheck/all/conanfile.py:32]:
```suggestion
            self.requires("z3/4.10.2")
```

**Comment by ericriff** [comment]
Can we include 
```
    def package_id(self):
        del self.info.settings.compiler
```
?

**Comment by uilianries** [comment]
> Can we include
> 
> ```
>     def package_id(self):
>         del self.info.settings.compiler
> ```
> 
> ?

Yes, please. Documented here: https://github.com/conan-io/conan-center-index/blob/master/docs/packaging_policy.md#settings

**Review by uilianries — COMMENTED** [review]
please, add to cpp_info, self.buildenv_info and self.runenv_info: https://docs.conan.io/en/latest/migrating_to_2.0/recipes.html#removed-self-env-info Only for CPPCHECK_HTMLREPORT. Path is automatically managed by VirtualRunEnv generator

**Comment by MartinDelille** [comment]
Blocked by https://github.com/conan-io/conan-center-index/pull/13186

**Comment by MartinDelille** [comment]
Replaced by https://github.com/conan-io/conan-center-index/pull/14149

**Pull request closed without merging**